File: 2022/day03.py
from data.day03 import SAMPLE_TXT, INPUT_TXT
from string import ascii_letters


def part1(txt: str) -> int:
    overlaps = [get_common_items(line).pop() for line in txt.splitlines()]
    return calculate_total_value(overlaps)


def part2(txt: str) -> int:
    overlaps = [set.intersection(*elf_group).pop() for elf_group in get_threebies(txt.splitlines())]
    return calculate_total_value(overlaps)


def get_common_items(content: str) -> set[str]:
    half1, half2 = content[:len(content)//2], content[len(content)//2:]
    overlaps = set(half1) & set(half2)
    return overlaps


def get_threebies(lines: list[str]) -> list[set[str]]:
    for elves in zip(lines[::3], lines[1::3], lines[2::3]):
        yield [set(elf) for elf in elves]


def calculate_total_value(overlaps: list[str]) -> int:
    return sum(1 + ascii_letters.index(overlap) for overlap in overlaps) # Neat Trick from Joel Grus + LiquidFun


##########################
if __name__ == '__main__':
    assert part1(SAMPLE_TXT) == 157
    print(part1(INPUT_TXT))
    assert part2(SAMPLE_TXT) == 70
    print(part2(INPUT_TXT))

File: 2022-kotlin/day03.kt
import day3_data.INPUT_TXT
import day3_data.SAMPLE_TXT


private fun part1(txt: String): Int {
    return txt.lines().map{it.splitInHalf()}.sumOf { calculateTotalValue(it)}
}

private fun part2(txt: String): Int {
    return txt.lines().byThree().sumOf { calculateTotalValue(it) }
}

private fun intersectingChar(elfBags: List<String>): Char{
    return elfBags.map{it.toSet()}.reduce{ sum, element -> sum.intersect(element)}.toCharArray()[0]
}

private fun calculateTotalValue(elfBags: List<String>): Int {
    val overlap = intersectingChar(elfBags)
    if (overlap in 'a'..'z') {
        return overlap.code - 96
    }
    return overlap.code - 38
}


private fun String.splitInHalf(): List<String>{
    return this.chunked(this.length / 2)
}


private fun List<String>.byThree(): List<List<String>>{
    return this.withIndex().groupBy{ (it.index / 3) }.values.map{it.map{it2 -> it2.value}}
}

fun main() {
    assert(part1(SAMPLE_TXT) == 157)
    println(part1(INPUT_TXT))
    assert(part2(SAMPLE_TXT) == 70)
    println(part2(INPUT_TXT))
}